rts/System/Platform/Threading.cpp

```diff
@@ -130,17 +134,47 @@ namespace Threading {
 		// This doesn't preclude systems from using separate unpinned threads, which the OS should logically try to
 		// move to under used resources, such as low-power cores for example.
 		#if defined(THREADPOOL)
-		uint32_t policy = pm.performanceCoreMask & (~pm.hyperThreadHighMask);
+		const uint32_t policy = pm.performanceCoreMask & (~pm.hyperThreadHighMask);
 		#else
 
 		/* Allow any core; keep it a "proper" mask though
 		 * since that has less risk of blowing up than 0 or 0xFF..FF */
-		uint32_t policy = pm.performanceCoreMask | pm.efficiencyCoreMask;
+		const uint32_t policy = pm.performanceCoreMask | pm.efficiencyCoreMask;
 		#endif
 
 		return policy;
 	}
 
+	uint32_t GetPreferredMainThreadMask() {
+		cpu_topology::ProcessorCaches pc = springproc::CPUID::GetInstance().GetProcessorCaches();
+
+		#if defined(THREADPOOL)
+		const uint32_t affinityMask = GetSystemAffinityMask();
+		auto preferredCache = std::find_if(pc.groupCaches.begin(), pc.groupCaches.end()
+			, [affinityMask](const auto& gc) -> bool { return !!(affinityMask & gc.groupMask); });
+		
+		static bool runOnce = false;
+		if (!runOnce) {
+			if (preferredCache != pc.groupCaches.end())
+				LOG("[Threading] Preferred performance cache mask is: 0x%08x (L3 sized: %dKB)", (*preferredCache).groupMask, (*preferredCache).cacheSizes[2]/1024);
+			else
+				LOG_L(L_WARNING, "[Threading] Failed to find a preferred performance cache mask");
+			runOnce = true;
+		}
+
+		const uint32_t policy = affinityMask
+			& ( (preferredCache != pc.groupCaches.end()) ? (*preferredCache).groupMask : 0xffffffff );
+		#else
+		/* Allow any core; keep it a "proper" mask though
+		 * since that has less risk of blowing up than 0 or 0xFF..FF */
+		cpu_topology::ProcessorMasks pm = springproc::CPUID::GetInstance().GetAvailableProcessorAffinityMask();
+		const uint32_t policy = pm.performanceCoreMask | pm.efficiencyCoreMask;
+		#endif
+
+		// Choose last logicial processor in the list.
```

Comment:
Are we choosing the last one on the list because there was `std::sort` earlier that put the largest-cache processor (i.e. the best choice) there?
```suggestion
		// Choose last logical processor in the list, they're sorted by size
```

Reply:
Updated.